### PR TITLE
Faster and simpler filter implementation

### LINKED ISSUE #10: 'in' (and '!in') filters fail with ~64k elements in set
There's a stack overflow due to all of the filter elements being passed as arguments. This bug is more important if #8 is fixed because that opens up use cases where this could be useful.

### PR DESCRIPTION
**Pull request by mourner**
Closes #10, ref #8. Alternative approach, still using eval.

Filters have the same performance, but filter creation is 40% faster, and there is 30% less code. 

It also makes `in` with a lot of items faster due to using `indexOf`, and it'll be easier to significantly improve further on this basis.

cc @jfirebaugh @dcervelli

### CODE REVIEW

**Comment by mourner** [comment]
Benchmarks indicate that it works much faster for simpler filters and on par with #9 on huge filters: 

```
# master
create 5-item filter: 1.174ms
filter 1 million times: 89.382ms
create 64000-item filter: 345.341ms
filter 1 million times: 726.251ms

# pr 9
create 5-item filter: 1.081ms
filter 1 million times: 89.812ms
create 64000-item filter: 290.508ms
gfilter 1 million times: 132.900ms

# simple 
create 5-item filter: 0.796ms
filter 1 million times: 30.522ms
create 64000-item filter: 28.999ms
filter 1 million times: 164.089ms
```

**Comment by mourner** [comment]
Disregard the 64k results, had a faulty benchmark. Here is a correct benchmark for 2k:

```
# master 
create 2000-item filter: 10.961ms
filter 1 million times: 20915.215ms

# pr 9
create 2000-item filter: 6.585ms
filter 1 million times: 215.851ms

# simple
create 2000-item filter: 0.747ms
filter 1 million times: 1414.595ms
```

So for filtering against a huge number of items, this branch is much faster than master but much slower than pr/9. I'll now try to implement binary search and this should change things dramatically though.

**Comment by mourner** [comment]
Added binary search! It turns on for more than 200 items (picked by testing running benchmarks). Results are great:

```
# pr 9
create 100000-item filter: 428.668ms
filter 1 million times: 499.984ms

# simple
create 100000-item filter: 71.494ms
filter 1 million times: 383.677ms
```

**Comment by dcervelli** [comment]
@mourner Awesome! Way simpler. Let me quickly add a benchmark on randomly generated strings as the filter items instead of numbers.

**Comment by dcervelli** [comment]
For random string data (which is closer to our use case) I get these results:

```
# pr 9
create 64000-item filter: 240.739ms
filter 1 million times: 757.678ms

# simple
create 64000-item filter: 161.316ms
filter 1 million times: 1879.914ms
```

Given the simpler code, I'm totally OK with this until it proves to be a bottleneck (which I doubt it will).

And for calibration: on my box for bench/big.js I get:

```
# simple
create 64000-item filter: 38.474ms
filter 1 million times: 324.140ms
```

Here's my benchmark:

``` javascript
'use strict';

var filter = require('../');

var N = 64000;

var arr = ['in', 'foo'];
for (var i = 0; i < N; i++) arr.push(i + " " + Math.random());

console.time('create ' + N + '-item filter');
var f = filter(arr);
console.timeEnd('create ' + N + '-item filter');

var feature = {properties: {foo: 0}};

console.time('filter 1 million times');
for (var i = 0; i < 1000000; i++) {
    feature.properties.foo = arr[Math.floor(Math.random() * N) + 2];
    f(feature);
}
console.timeEnd('filter 1 million times');

```

Thanks!

File: test.js [hunk 1]
```diff
@@ -294,7 +294,7 @@ test('!in, $type', function(t) {
 
 test('any', function(t) {
     var f1 = filter(['any']);
-    t.equal(f1({properties: {foo: 1}}), false);
+    t.equal(f1({properties: {foo: 1}}), true);
```
Comment (jfirebaugh) [inline test.js]: This shouldn't change; returning `false` for an empty `any` is standard mathematical / functional programming behavior.

File: index.js [hunk 2]
```diff
@@ -70,9 +12,51 @@ function truth() {
...
+        op === 'all' ? filter.slice(1).map(compile).join('&&') :
+        op === 'none' ? '!(' + filter.slice(1).map(compile).join('||') + ')' :
+        op === 'in' ? compileIn(filter[1], filter.slice(2)) :
+        op === '!in' ? '!(' + compileIn(filter[1], filter.slice(2)) + ')' :
+        'true';
+    return '(' + str + ')';
+}
+
+function valueExpr(key) {
+    return key === '$type' ? 'f.type' : '(f.properties || {})[' + JSON.stringify(key) + ']';
```
Comment (jfirebaugh) [inline index.js]: Previous implementation cached `f.type` and `f.properties || f.tags || {}` in local variables.

* Does caching affect performance?
* Do we need to keep supporting `f.tags`?

Comment (mourner) [inline index.js]: > Does caching affect performance?

As far as I remember, doesn't, but I can readd and make sure now.

> Do we need to keep supporting `f.tags`?

I don't know why they were supported in the first place. When we removed filtering code from gl-js (https://github.com/mapbox/mapbox-gl-js/commit/96013eb9134c50262a61c4a7a4d08d5fa869cddf), it was just `f.properties`, but the first commit here has `f.tags`. Maybe @camilleanne knows?

File: test.js (hunk 1 above)
Comment (mourner) [inline test.js]: Didn't know that, it's not very intuitive. Will fix.

File: index.js (hunk 2 above)
Comment (camilleanne) [inline index.js]: @mourner Hm, I didn't modify this code, I just ported it, so I'm in the dark.

Comment (mourner) [inline index.js]: > Does caching affect performance?

Checked, it doesn't.

**Comment by mourner** [comment]
@dcervelli great! I think it's still slower than it could be due to the fact that you have to evaluate the whole huge array on every filter function call. For future improvements, we can make an attempt at creating a closure that would contain all the big arrays referenced in filters so that the function call itself remains lightweight.

**Pull request merged**